### PR TITLE
use v2 streams

### PR DESCRIPTION
**Pull request by rynomad**

### CODE REVIEW

File: packages/gateway-client/src/worker/p2p-client/stream-factory.ts
Comment (JoshuaCWebDeveloper) [inline packages/gateway-client/src/worker/p2p-client/stream-factory.ts:226]: I'm not sure I like coupling a specific protocol with different stream logic this tightly. I'd like to see a separation between the `samizdapp-proxy` protocol, the concept of a stream pool, and the concept of a request stream?

Why not introduce the pooling concept to `getStream()`? You could add a second optional options parameter to it with a `pooled = false` property.

The above would result in:
 - `getStream()` with `protocol` and `options` (pooled = false) params
 - `getProxyStream()` is renamed to `getRequestStream()` that takes same two params
   - Both of these params are passed to `new RequestStream()` (which, in line of the above comment, in turn calls `getStream()`)

If you want to introduce *some* coupling between the proxy protocol and the request stream logic, there are some additional patterns you could implement.

File: packages/gateway-client/src/worker/p2p-client/stream-factory.ts
Comment (JoshuaCWebDeveloper) [inline packages/gateway-client/src/worker/p2p-client/stream-factory.ts:245]: Why do we need this method instead of just calling `close()` on the stream directly?

File: packages/gateway-client/src/worker/p2p-client/stream-factory.ts
Comment (JoshuaCWebDeveloper) [inline packages/gateway-client/src/worker/p2p-client/stream-factory.ts:241]: Ideally there could just be a `release()` method on the returned stream that is a no-op if the stream wasn't pooled to begin with. But if that is too complex or too much to implement, this is probably fine.

File: packages/gateway-client/src/worker/p2p-fetch/p2p-request.ts
Comment (JoshuaCWebDeveloper) [inline packages/gateway-client/src/worker/p2p-fetch/p2p-request.ts:58]: We can probably get rid of this commented code before merging, yeah? Or is there a specific reason for keeping it?

**Review by JoshuaCWebDeveloper — CHANGES_REQUESTED** [review]
I really like the introduction of pooling logic, that should make our use of streams much more efficient!

I'm also glad to see us introducing a "request stream" pattern and generalizing some of the fetch logic. In fact, this PR probably allowed you to go a lot farther in terms of generalizing the p2p request attempt pattern as well, but that can be done more easily in a future PR now that this PR generalizes the duplex logic.

I actually see some opportunities to make the logic even a bit more general and reusable. Doing this would also make the concepts a bit easier to understand, so I've requested some refactoring to separate out a "request stream", a "pooled stream", and specific protocols.

Also, the reusable logic in the request stream was pretty difficult to understand, so I've requested some more modest changes to the logic which should make it read a bit easier.

Happy to talk through any of this, nice job!!

File: packages/gateway-client/src/worker/p2p-client/stream-factory.ts
Comment (rynomad) [reply packages/gateway-client/src/worker/p2p-client/stream-factory.ts:49]: This is doable, and yes, the reason for the `open` sink generator is that calling sink is only allowed once per stream.

File: packages/gateway-client/src/worker/p2p-fetch/p2p-fetch-request.ts
Comment (rynomad) [reply packages/gateway-client/src/worker/p2p-fetch/p2p-fetch-request.ts:241]: yeah I'll take that out. for a second I thought I needed it.

File: packages/gateway-client/src/worker/p2p-client/stream-factory.ts
Comment (JoshuaCWebDeveloper) [inline packages/gateway-client/src/worker/p2p-client/stream-factory.ts:9]: I'd actually prefer that we copy the class instead of re-using it, just because it is pretty simple and I don't want to introduce "util" type stuff (i.e. the way that we're also copying `waitFor` all over the place).

File: packages/gateway-client/src/worker/p2p-client/stream-factory.ts
Comment (JoshuaCWebDeveloper) [inline packages/gateway-client/src/worker/p2p-client/stream-factory.ts:32]: Nice.

File: packages/gateway-client/src/worker/p2p-client/stream-factory.ts
Comment (JoshuaCWebDeveloper) [inline packages/gateway-client/src/worker/p2p-client/stream-factory.ts:39]: Just wanted to highlight that you're now yielding a `Buffer` instead of a `UInt8Array`

File: packages/gateway-client/src/worker/p2p-fetch/p2p-fetch-request.ts
Comment (JoshuaCWebDeveloper) [inline packages/gateway-client/src/worker/p2p-fetch/p2p-fetch-request.ts:286]:
```suggestion
        // create a new response to return	
        const response = new Response(resp.body, resp.json.res);
```

**Review by JoshuaCWebDeveloper — APPROVED** [review]
Approving with a couple of last requests.

File: packages/gateway-client/src/worker/p2p-client/stream-factory.ts
Comment (rynomad) [reply packages/gateway-client/src/worker/p2p-client/stream-factory.ts:39]: yeah, that's intentional, and Buffer inherits from Uint8Array so anything that depends on that api will still work.

**Pull request merged**
